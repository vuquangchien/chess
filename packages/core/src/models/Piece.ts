import {PieceColor} from './PieceColor';
import {PieceType} from './PieceType';
export type FenName =
  | 'p'
  | 'r'
  | 'n'
  | 'b'
  | 'q'
  | 'k'
  | 'P'
  | 'R'
  | 'N'
  | 'B'
  | 'Q'
  | 'K';
export const PieceTypes: {[key: string]: PieceType} = {
  PAWN: 'pawn',
  ROOK: 'rook',
  KNIGHT: 'knight',
  BISHOP: 'bishop',
  QUEEN: 'queen',
  KING: 'king',
};
export const PieceColors: {[key: string]: PieceColor} = {
  WHITE: 'white',
  BLACK: 'black',
};
export class Piece {
  public color: PieceColor;
  public type: PieceType;
  private isMoved = false;

  constructor(type: PieceType, color: PieceColor) {
    this.color = color;
    this.type = type;
  }

  static fromName(name: FenName): Piece {
    const color =
      name === name.toUpperCase() ? PieceColors.WHITE : PieceColors.BLACK;
    const type = name.toLowerCase();
    const pieceType = {
      p: PieceTypes.PAWN,
      r: PieceTypes.ROOK,
      n: PieceTypes.KNIGHT,
      b: PieceTypes.BISHOP,
      q: PieceTypes.QUEEN,
      k: PieceTypes.KING,
    }[type];
    if (!pieceType) throw new Error(`Unknown piece type: ${type}`);
    return new Piece(pieceType, color);
  }

  public getColor(): PieceColor {
    return this.color;
  }

  public isPieceMoved(): boolean {
    return this.isMoved;
  }
  public setPieceMoved(): void {
    this.isMoved = true;
  }
  public getType(): PieceType {
    return this.type;
  }
  public getIcon(): string {
    const icons = {
      rook: {
        white: '♖',
        black: '♜',
      },
      knight: {
        white: '♘',
        black: '♞',
      },
      bishop: {
        white: '♗',
        black: '♝',
      },
      queen: {
        white: '♕',
        black: '♛',
      },
      king: {
        white: '♔',
        black: '♚',
      },
      pawn: {
        white: '♙',
        black: '♟',
      },
    };
    return icons[this.type][this.color];
  }
}
