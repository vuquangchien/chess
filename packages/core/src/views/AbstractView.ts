import {IView} from '../framework/IView';
import {IEngine} from './IEngine';

export abstract class AbstractView<T> implements IView<T> {
  engine!: IEngine;
  children: AbstractView<unknown>[] = [];
  controller: T;

  public constructor(controller: T) {
    this.children = [];
    this.controller = controller;
    this.init();
  }
  init(): void {}

  setEngine(engine: IEngine): void {
    this.engine = engine;
    this.children.forEach(child => child.setEngine(engine));
  }
  addChild(child: AbstractView<unknown>): void {
    this.children.push(child);
  }

  addChildren(children: AbstractView<unknown>[]): void {
    this.children.push(...children);
  }

  renderChildren(): string {
    let result = '';
    for (const child of this.children) {
      result += child.render();
    }
    return result;
  }

  abstract render(): string;
}
