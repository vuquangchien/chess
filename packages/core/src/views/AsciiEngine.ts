import {IEngine} from './IEngine';

export class AsciiEngine implements IEngine {
  renderBox(options: {background: string; text: string}): string {
    const {background, text} = options;
    return (background === '#000' ? '■' : '□') + text;
  }
}
