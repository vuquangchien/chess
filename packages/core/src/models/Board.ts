import {Cell} from './Cell';
import {CellIterator} from './CellIterator';
import * as _ from 'lodash';
import {Move} from './moves/Move';
import {MoveFactory} from './moves/MoveFactory';
import {PieceColors} from './Piece';
import {PieceColor} from './PieceColor';
import {PolicyFactory} from './policies/PolicyFactory';

export class Board {
  public cells: Cell[][];
  public moves: Move[] = [];
  private player = PieceColors.WHITE;
  constructor() {
    // const cellIndex = 0;
    const files: Cell[][] = [];
    for (let rowIndex = 0; rowIndex < 8; rowIndex++) {
      const file: Cell[] = [];
      for (let colIndex = 0; colIndex < 8; colIndex++) {
        const cell = new Cell(colIndex, rowIndex, null);
        file.push(cell);
      }
      files.push(file);
    }
    this.cells = files;
  }

  public getByCellName(name: string): Cell {
    try {
      const file = name.charCodeAt(0) - 97;
      const rank = 8 - parseInt(name.charAt(1), 10);
      return this.cells[rank][file];
    } catch (e) {
      throw new Error(`Invalid cell name ${name}`);
    }
  }

  public assertMove(move: Move): void {
    const movePolicies = PolicyFactory.getPolicies(this, move);
    movePolicies.forEach(policy => {
      const error = policy.isMoveValid(move);
      if (error) {
        throw new Error(error);
      }
    });
  }

  public movePiece(from: Cell, to: Cell): void {
    const move = MoveFactory.createMove(from, to);
    this.assertMove(move);
    this.moves.push(move);
    to.setPiece(from.piece);
    from.piece!.setPieceMoved();
    from.setPiece(null);
    this.player =
      this.player === PieceColors.WHITE ? PieceColors.BLACK : PieceColors.WHITE;
  }
  public movePieceByCellName(from: string, to: string): void {
    const fromCell = this.getByCellName(from);
    const toCell = this.getByCellName(to);
    this.movePiece(fromCell, toCell);
  }
  public getCellIterator(): CellIterator {
    return new CellIterator(_.flatten(this.cells));
  }
  public highlightPossibleMoves(cell: Cell) {
    const possibleMoves = this.getPossibleMoves(cell);
    possibleMoves.forEach(possibleMove => {
      possibleMove.highlight();
    });
  }
  public getPossibleMoves(fromCell: Cell): Cell[] {
    const possibleMoves: Cell[] = [];
    const it = this.getCellIterator();
    while (it.hasNext()) {
      const toCell = it.next();
      try {
        const move = MoveFactory.createMove(fromCell, toCell);
        this.assertMove(move);
        possibleMoves.push(toCell);
      } catch (e) {
        // console.log(e);
      }
    }
    return possibleMoves;
  }
  public getAllPossibleMoves(): Move[] {
    const possibleMoves: Move[] = [];
    const it = this.getCellIterator();
    while (it.hasNext()) {
      const fromCell = it.next();
      if (!fromCell.piece) {
        continue;
      }
      const moves = this.getPossibleMoves(fromCell);
      moves.forEach(toCell => {
        possibleMoves.push(MoveFactory.createMove(fromCell, toCell));
      });
    }
    return possibleMoves;
  }
  resetHighlight() {
    const it = this.getCellIterator();
    while (it.hasNext()) {
      it.next().unHighlight();
    }
  }

  getCurrentPlayer() {
    return this.player;
  }

  setCurrentPlayer(player: PieceColor) {
    this.player = player;
  }

  clone() {
    return _.cloneDeep(this);
  }
}
