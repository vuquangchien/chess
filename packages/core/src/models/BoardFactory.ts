import {Board} from './Board';
import {FenRecord} from './fen/FenRecord';

export class BoardFactory {
  public static createEmptyBoard(): Board {
    return new Board();
  }
  public static createFromFen(fen: string): Board {
    const board = new Board();
    const fenRecord = new FenRecord(fen);
    const cells = fenRecord.piecePlacement.getCells();
    cells.forEach(cell => {
      board.getByCellName(cell.name).setPiece(cell.piece);
    });
    board.setCurrentPlayer(fenRecord.activeColor);
    return board;
  }
  public static createBoardWithPieces(): Board {
    return BoardFactory.createFromFen(
      'rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w - - 0 1'
    );
  }
}
