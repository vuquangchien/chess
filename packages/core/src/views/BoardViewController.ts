import {Board} from '../models/Board';

export class BoardViewController {
  private board: Board;

  constructor(board: Board) {
    this.board = board;
  }

  public getBoard(): Board {
    return this.board;
  }
}
