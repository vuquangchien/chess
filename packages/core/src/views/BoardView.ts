import {CellView} from './CellView';
import {CellViewController} from './CellViewController';

import {BoardViewController} from './BoardViewController';
import {RowEl, TextEl} from './RowEl';
import {AbstractView} from './AbstractView';

export class BoardView extends AbstractView<BoardViewController> {
  public init(): void {
    this.addChild(new RowEl());
    const headerRow = new RowEl();
    headerRow.addChildren(
      '   a b c d e f g h'
        .split(' ')
        .map(el => new TextEl({text: el + '', row: 0, column: 0}))
    );
    this.addChild(headerRow);
    const board = this.controller.getBoard();
    const iterator = board.getRankIterator();
    while (iterator.hasNext()) {
      const row = new RowEl();
      row.addChild(new TextEl({text: `${7 - iterator.currentIndex()} `}));
      const rank = iterator.next();
      const it = rank.getIterator();
      while (it.hasNext()) {
        const cell = it.next();
        const vc = new CellViewController(cell);
        const cellView = new CellView(vc);
        row.addChild(cellView);
      }
      this.addChild(row);
    }
  }

  public render(): string {
    return this.renderChildren();
  }
}
