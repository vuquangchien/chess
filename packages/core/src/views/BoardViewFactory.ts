import {Board} from '../models/Board';
import {BoardViewController} from './BoardViewController';
import {BoardView} from './BoardView';

export class BoardViewFactory {
  public static fromBoard(board: Board): BoardView {
    return new BoardView(new BoardViewController(board));
  }
}
