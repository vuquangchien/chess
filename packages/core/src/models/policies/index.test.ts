import {Board} from '../Board';
// @ts-ignore
import {expect} from '@jest/globals';
import {BoardFactory} from '../BoardFactory';
import {BoardView} from '../../views/BoardView';
import {BoardViewController} from '../../views/BoardViewController';
import {BoardViewFactory} from '../../views/BoardViewFactory';
import {Piece} from '../Piece';

describe('policies', () => {
  it('not allow to move empty piece', () => {
    const board = new Board();
    expect(() => {
      board.movePieceByCellName('a5', 'a4');
    }).toThrowError();
  });
  it('not allow to move piece to the same cell', () => {
    const board = BoardFactory.createBoardWithPieces();
    expect(() => {
      board.movePieceByCellName('a7', 'a7');
    }).toThrowError();
  });
  it('not allow to move piece to the cell with the same color', () => {
    const board = BoardFactory.createBoardWithPieces();
    console.log(new BoardView(new BoardViewController(board)).render());
    expect(() => {
      board.movePieceByCellName('a8', 'a7');
    }).toThrowError();
  });
  it('not allow to white or black to move twice in a row', () => {
    const board = BoardFactory.createBoardWithPieces();
    expect(() => {
      board.movePieceByCellName('a7', 'a6');
      board.movePieceByCellName('a6', 'a5');
    }).toThrowError();
  });
  it('should allow to take', () => {
    const board = BoardFactory.createFromFen(
      'rnbqkbnr/pppp1ppp/8/4p3/3P4/8/PPP1PPPP/RNBQKBNR b KQkq e6 0 1'
    );
    console.log(BoardViewFactory.fromBoard(board).render());
    expect(() => {
      board.movePieceByCellName('e5', 'd4');
    }).not.toThrowError();
  });
  // highlight tests
  describe('highlight', () => {
    // highlight cell
    it('should highlight cell', () => {
      const board = BoardFactory.createBoardWithPieces();
      const cell = board.getByCellName('a8');
      cell.highlight();
      expect(cell.isHighlighted()).toBeTruthy();
    });
    // unhighlight cell
    it('should unhighlight cell', () => {
      const board = BoardFactory.createBoardWithPieces();
      const cell = board.getByCellName('a8');
      cell.highlight();
      cell.unHighlight();
      expect(cell.isHighlighted()).toBeFalsy();
    });
    // reset highlight
    it('should reset highlight', () => {
      const board = BoardFactory.createBoardWithPieces();
      const cell = board.getByCellName('a8');
      cell.highlight();
      board.resetHighlight();
      expect(cell.isHighlighted()).toBeFalsy();
    });
  });

  describe('getPossibleMoves', () => {
    // king
    describe('king', () => {
      it('should get possible moves for king', () => {
        const board = new Board();
        const king = Piece.fromName('K');
        const cell = board.getByCellName('d4');
        cell.setPiece(king);

        board.highlightPossibleMoves(cell);
        const view = BoardViewFactory.fromBoard(board).render();
        expect(view).toMatchSnapshot();
        const possibleMoves = board.getPossibleMoves(cell);

        expect(possibleMoves.length).toEqual(8);
      });
    });
  });
  // queen
  describe('queen', () => {
    it('should get possible moves for queen', () => {
      const board = new Board();
      const queen = Piece.fromName('Q');
      const cell = board.getByCellName('d4');
      cell.setPiece(queen);
      board.highlightPossibleMoves(cell);
      const view = BoardViewFactory.fromBoard(board).render();
      expect(view).toMatchSnapshot();
      const possibleMoves = board.getPossibleMoves(cell);

      expect(possibleMoves.length).toEqual(27);
    });
  });
  // rook
  describe('rook', () => {
    it('should get possible moves for rook', () => {
      const board = new Board();
      const rook = Piece.fromName('R');
      const cell = board.getByCellName('d4');
      cell.setPiece(rook);
      board.highlightPossibleMoves(cell);
      const view = BoardViewFactory.fromBoard(board).render();
      expect(view).toMatchSnapshot();
      const possibleMoves = board.getPossibleMoves(cell);

      expect(possibleMoves.length).toEqual(14);
    });
  });
  // bishop
  describe('bishop', () => {
    it('should get possible moves for bishop', () => {
      const board = new Board();
      const bishop = Piece.fromName('B');
      const cell = board.getByCellName('d4');
      cell.setPiece(bishop);
      board.highlightPossibleMoves(cell);
      const view = BoardViewFactory.fromBoard(board).render();
      expect(view).toMatchSnapshot();
      const possibleMoves = board.getPossibleMoves(cell);

      expect(possibleMoves.length).toEqual(13);
    });
  });
  // knight
  describe('knight', () => {
    it('should get possible moves for knight', () => {
      const board = new Board();
      const knight = Piece.fromName('N');
      const cell = board.getByCellName('d4');
      cell.setPiece(knight);
      board.highlightPossibleMoves(cell);
      const view = BoardViewFactory.fromBoard(board).render();
      expect(view).toMatchSnapshot();
      const possibleMoves = board.getPossibleMoves(cell);

      expect(possibleMoves.length).toEqual(8);
    });
  });
  // pawn
  describe('pawn', () => {
    it('should get possible moves for pawn', () => {
      const board = new Board();
      const pawn = Piece.fromName('P');
      pawn.setPieceMoved();
      const cell = board.getByCellName('d4');
      cell.setPiece(pawn);
      board.highlightPossibleMoves(cell);
      const view = BoardViewFactory.fromBoard(board).render();
      expect(view).toMatchSnapshot();
      const possibleMoves = board.getPossibleMoves(cell);

      expect(possibleMoves.length).toEqual(1);
    });
    // pawn first move
    it('should get possible moves for pawn on first move', () => {
      const board = new Board();
      const pawn = Piece.fromName('P');
      const cell = board.getByCellName('d2');
      cell.setPiece(pawn);
      board.highlightPossibleMoves(cell);
      const view = BoardViewFactory.fromBoard(board).render();
      expect(view).toMatchSnapshot();
      const possibleMoves = board.getPossibleMoves(cell);
      expect(possibleMoves.length).toEqual(2);
    });
  });
});
