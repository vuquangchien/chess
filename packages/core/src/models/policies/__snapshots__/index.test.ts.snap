// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`policies bishop should get possible moves for bishop 1`] = `
"
□□□□a□b□c□d□e□f□g□h
□8 □ □ □ □ □ □ □ □ 
□7 □ □ □ □ □ □ □ □ 
□6 □ □ □ □ □ □ □ □ 
□5 □ □ □ □ □ □ □ □ 
□4 □ □ □ □♗□ □ □ □ 
□3 □ □ □ □ □ □ □ □ 
□2 □ □ □ □ □ □ □ □ 
□1 □ □ □ □ □ □ □ □ 
"
`;

exports[`policies getPossibleMoves king should get possible moves for king 1`] = `
"
□□□□a□b□c□d□e□f□g□h
□8 □ □ □ □ □ □ □ □ 
□7 □ □ □ □ □ □ □ □ 
□6 □ □ □ □ □ □ □ □ 
□5 □ □ □ □ □ □ □ □ 
□4 □ □ □ □♔□ □ □ □ 
□3 □ □ □ □ □ □ □ □ 
□2 □ □ □ □ □ □ □ □ 
□1 □ □ □ □ □ □ □ □ 
"
`;

exports[`policies knight should get possible moves for knight 1`] = `
"
□□□□a□b□c□d□e□f□g□h
□8 □ □ □ □ □ □ □ □ 
□7 □ □ □ □ □ □ □ □ 
□6 □ □ □ □ □ □ □ □ 
□5 □ □ □ □ □ □ □ □ 
□4 □ □ □ □♘□ □ □ □ 
□3 □ □ □ □ □ □ □ □ 
□2 □ □ □ □ □ □ □ □ 
□1 □ □ □ □ □ □ □ □ 
"
`;

exports[`policies pawn should get possible moves for pawn 1`] = `
"
□□□□a□b□c□d□e□f□g□h
□8 □ □ □ □ □ □ □ □ 
□7 □ □ □ □ □ □ □ □ 
□6 □ □ □ □ □ □ □ □ 
□5 □ □ □ □ □ □ □ □ 
□4 □ □ □ □♙□ □ □ □ 
□3 □ □ □ □ □ □ □ □ 
□2 □ □ □ □ □ □ □ □ 
□1 □ □ □ □ □ □ □ □ 
"
`;

exports[`policies pawn should get possible moves for pawn on first move 1`] = `
"
□□□□a□b□c□d□e□f□g□h
□8 □ □ □ □ □ □ □ □ 
□7 □ □ □ □ □ □ □ □ 
□6 □ □ □ □ □ □ □ □ 
□5 □ □ □ □ □ □ □ □ 
□4 □ □ □ □ □ □ □ □ 
□3 □ □ □ □ □ □ □ □ 
□2 □ □ □ □♙□ □ □ □ 
□1 □ □ □ □ □ □ □ □ 
"
`;

exports[`policies queen should get possible moves for queen 1`] = `
"
□□□□a□b□c□d□e□f□g□h
□8 □ □ □ □ □ □ □ □ 
□7 □ □ □ □ □ □ □ □ 
□6 □ □ □ □ □ □ □ □ 
□5 □ □ □ □ □ □ □ □ 
□4 □ □ □ □♕□ □ □ □ 
□3 □ □ □ □ □ □ □ □ 
□2 □ □ □ □ □ □ □ □ 
□1 □ □ □ □ □ □ □ □ 
"
`;

exports[`policies rook should get possible moves for rook 1`] = `
"
□□□□a□b□c□d□e□f□g□h
□8 □ □ □ □ □ □ □ □ 
□7 □ □ □ □ □ □ □ □ 
□6 □ □ □ □ □ □ □ □ 
□5 □ □ □ □ □ □ □ □ 
□4 □ □ □ □♖□ □ □ □ 
□3 □ □ □ □ □ □ □ □ 
□2 □ □ □ □ □ □ □ □ 
□1 □ □ □ □ □ □ □ □ 
"
`;
